sandbox/modules/custom-codecs/src/main/java/org/opensearch/index/codec/customcodec/ZstdCompressionMode.java

```diff
@@ -0,0 +1,200 @@
+            if (len == 0) {
+                out.writeVInt(0);
+                return;
+            }
+            final int maxCompressedLength = (int) Zstd.compressBound(len);
+            compressedBuffer = ArrayUtil.grow(compressedBuffer, maxCompressedLength);
+
+            int compressedSize = cctx.compressByteArray(compressedBuffer, 0, compressedBuffer.length, bytes, off, len);
+
+            out.writeVInt(compressedSize);
+            out.writeBytes(compressedBuffer, compressedSize);
+        }
+
+        private void compress(byte[] bytes, int off, int len, DataOutput out) throws IOException {
+            final int dictLength = len / (NUM_SUB_BLOCKS * DICT_SIZE_FACTOR);
+            final int blockLength = (len - dictLength + NUM_SUB_BLOCKS - 1) / NUM_SUB_BLOCKS;
+            out.writeVInt(dictLength);
+            out.writeVInt(blockLength);
+
+            final int end = off + len;
+
+            try (ZstdCompressCtx cctx = new ZstdCompressCtx()) {
+                cctx.setLevel(this.compressionLevel);
+
+                // dictionary compression first
+                doCompress(bytes, off, dictLength, cctx, out);
+                cctx.loadDict(new ZstdDictCompress(bytes, off, dictLength, this.compressionLevel));
+
+                for (int start = off + dictLength; start < end; start += blockLength) {
+                    int l = Math.min(blockLength, off + len - start);
+                    doCompress(bytes, start, l, cctx, out);
+                }
+            }
+        }
+
+        @Override
+        public void compress(ByteBuffersDataInput buffersInput, DataOutput out) throws IOException {
+            final int len = (int) buffersInput.size();
+            byte[] bytes = new byte[len];
+            buffersInput.readBytes(bytes, 0, len);
+            compress(bytes, 0, len, out);
+        }
+
+    }
+
+    /** zstandard decompressor */
+    private static final class ZSTDDecompressor extends Decompressor {
+
+        private byte[] compressedBuffer;
+
+        /** default decompressor */
+        public ZSTDDecompressor() {
+            compressedBuffer = BytesRef.EMPTY_BYTES;
+        }
+
+        /*resuable decompress function*/
+        private void doDecompress(DataInput in, ZstdDecompressCtx dctx, BytesRef bytes, int decompressedLen) throws IOException {
+            final int compressedLength = in.readVInt();
+            if (compressedLength == 0) {
+                return;
+            }
+
+            compressedBuffer = ArrayUtil.grow(compressedBuffer, compressedLength);
+            in.readBytes(compressedBuffer, 0, compressedLength);
+
+            bytes.bytes = ArrayUtil.grow(bytes.bytes, bytes.length + decompressedLen);
+            int uncompressed = dctx.decompressByteArray(bytes.bytes, bytes.length, decompressedLen, compressedBuffer, 0, compressedLength);
+
+            if (decompressedLen != uncompressed) {
+                throw new IllegalStateException(decompressedLen + " " + uncompressed);
```

Comment:
We should probably throw a specific exception here so that a useful error message bubbles up.

Reply:
I have addressed this in the latest commit.